### PR TITLE
Add security group participant message handling to net_plugin

### PR DESCRIPTION
**Pull request by rusty-block-one**
## Change Description
Added logic to net_plugin_impl and connection classes to support security group in privacy features. 

## Change Type
**Select *ONE*:**
- [ ] Documentation

- [ ] Stability bug fix

- [x] Other

- [ ] Other - special case

## Testing Changes
**Select *ANY* that apply:**
- [ ] New Tests

- [ ] Existing Tests

- [ ] Test Framework

- [ ] CI System

- [ ] Other

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

**Comment by rusty-block-one** [comment]
@brianjohnson5972 the PR is ready for review

**Comment by heifner** [comment]
Can we use descriptive names for PRs please.

**Comment by dimas1185** [comment]
FYI, EPE is private jira reference while PRs are public. Please substitute EPE with reasonable description.

**Comment by rusty-block-one** [comment]
@heifner @dimas1185 Changed name to descriptive name

File: plugins/net_plugin/net_plugin.cpp [hunk 1]
```diff
@@ -3919,6 +4013,16 @@ namespace eosio {
       if( my->find_connection( host ) )
          return "already connected";
 
+      /** @todo when making a shared tls connection, need to use the ctor that takes
```
Comment (brianjohnson5972) [inline plugins/net_plugin/net_plugin.cpp:4016]: Since setting the participant name has to be handled after the certificate handshake, just remove these comments

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -5,7 +5,7 @@ include_directories( "${CMAKE_SOURCE_DIR}/plugins/wallet_plugin/include" )
 file(GLOB UNIT_TESTS "*.cpp")
 
 add_executable( plugin_test ${UNIT_TESTS} )
-target_link_libraries( plugin_test eosio_testing eosio_chain chainbase chain_plugin wallet_plugin fc ${PLATFORM_SPECIFIC_LIBS} )
+target_link_libraries( plugin_test eosio_testing eosio_chain chainbase chain_plugin wallet_plugin fc net_plugin ${PLATFORM_SPECIFIC_LIBS} )
```
Comment (brianjohnson5972) [inline tests/CMakeLists.txt:8]: Why did you need to add this?

File: plugins/net_plugin/net_plugin.cpp [hunk 3]
```diff
@@ -789,6 +807,29 @@ namespace eosio {
             ( "_lport", local_endpoint_port );
          return mvo;
       }
+
+      // security group management
+      //
+   private:
+      std::optional<chain::account_name> participant_name_;
+      std::atomic<bool> participating_{false};
```
Comment (brianjohnson5972) [inline plugins/net_plugin/net_plugin.cpp:815]: default to true, the security_group changing logic and logic for TLS connections should handle setting it false when it needs to be

File: plugins/net_plugin/net_plugin.cpp [hunk 4]
```diff
@@ -789,6 +807,29 @@ namespace eosio {
...
+
+   public:
+      /** @brief  Set/change the participant name for the connection
+       *
+       *  @param name   The name of the participant
+       */
+      void participant_name(chain::account_name name) { participant_name_ = name; }
```
Comment (brianjohnson5972) [inline plugins/net_plugin/net_plugin.cpp:822]: The TLS handling for this should be in the connection, so leave off the setter

**Review by brianjohnson5972 — CHANGES_REQUESTED** [review]

File: plugins/net_plugin/net_plugin.cpp (hunk 1 above)
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:4016]: Removed comment

File: plugins/net_plugin/net_plugin.cpp (hunk 4 above)
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:822]: Removed setter

File: plugins/net_plugin/net_plugin.cpp (hunk 3 above)
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:815]: Changed default value

File: plugins/net_plugin/net_plugin.cpp [hunk 5]
```diff
@@ -789,6 +807,29 @@ namespace eosio {
...
+      /** @brief  Returns the optional name of the participant */
+      auto participant_name() { return participant_name_; }
+      /** returns true if connection is in the security group */
+      bool is_participating() const { return participating_.load(std::memory_order_relaxed); }
+       /** returns false if the connection is not in the security group */
+      bool is_syncing() const { return !participating_.load(std::memory_order_relaxed); }
+      /**   @brief    Flag the tls connection as participating in block production */
+      void now_participating() { participating_.store(true, std::memory_order_relaxed); }
```
[the remark this answers is not in the document]
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:830]: changed to use set_participating

File: plugins/net_plugin/net_plugin.cpp [hunk 6]
```diff
@@ -789,6 +807,29 @@ namespace eosio {
...
+   public:
+      /** @brief  Set/change the participant name for the connection
+       *
+       *  @param name   The name of the participant
+       */
+      void participant_name(chain::account_name name) { participant_name_ = name; }
+      /** @brief  Returns the optional name of the participant */
+      auto participant_name() { return participant_name_; }
+      /** returns true if connection is in the security group */
+      bool is_participating() const { return participating_.load(std::memory_order_relaxed); }
```
[the remark this answers is not in the document]
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:826]: The default for atomic is memory_order_seq_cst. I assume that all we want to do is change the value of the flag and not hold up any threads simultaneously accessing this value. With the default, the compiler may generate could that could temporarily block a thread in order to make sure there are no visible side effects.

File: plugins/net_plugin/net_plugin.cpp [hunk 7]
```diff
@@ -336,6 +339,17 @@ namespace eosio {
       constexpr static uint16_t to_protocol_version(uint16_t v);
 
       connection_ptr find_connection(const string& host)const; // must call with held mutex
+      /** private security group management */
+      security_group_manager security_group;
+      /** \brief  Update security group information
+       *
+       * Retrieves the security group information from the indicated block and
+       * updates the cached security group information. If a participant is
+       * removed from the cached security group, the TLS connection is closed.
```
[the remark this answers is not in the document]
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:348]: updated comment

File: plugins/net_plugin/net_plugin.cpp [hunk 8]
```diff
@@ -336,6 +339,17 @@ namespace eosio {
       constexpr static uint16_t to_protocol_version(uint16_t v);
 
       connection_ptr find_connection(const string& host)const; // must call with held mutex
+      /** private security group management */
+      security_group_manager security_group;
```
[the remark this answers is not in the document]
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:343]: moved

File: plugins/net_plugin/net_plugin.cpp [hunk 9]
```diff
@@ -1435,6 +1476,18 @@ namespace eosio {
 
    void connection::enqueue( const net_message& m ) {
       verify_strand_in_this_thread( strand, __func__, __LINE__ );
+      // for tls connections, when the connection is not in the security group
+      // certain message types will not be transmitted
+      if(participant_name_ && !participating_) {
```
[the remark this answers is not in the document]
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:1481]: changed now to just check _participating (by calling is_participating)

File: plugins/net_plugin/net_plugin.cpp [hunk 10]
```diff
@@ -256,6 +258,7 @@ namespace eosio {
 
       mutable std::shared_mutex             connections_mtx;
       std::set< connection_ptr >            connections;     // todo: switch to a thread safe container to avoid big mutex over complete collection
+      std::unordered_multimap<chain::account_name, connection_ptr> particiant_connections;
```
[the remark this answers is not in the document]
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:261]: removed

File: plugins/net_plugin/net_plugin.cpp [hunk 11]
```diff
@@ -789,6 +807,29 @@ namespace eosio {
...
+       *
+       *  @param name   The name of the participant
+       */
+      void participant_name(chain::account_name name) { participant_name_ = name; }
+      /** @brief  Returns the optional name of the participant */
+      auto participant_name() { return participant_name_; }
+      /** returns true if connection is in the security group */
+      bool is_participating() const { return participating_.load(std::memory_order_relaxed); }
+       /** returns false if the connection is not in the security group */
+      bool is_syncing() const { return !participating_.load(std::memory_order_relaxed); }
```
[the remark this answers is not in the document]
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:828]: changed to use is_participating

File: plugins/net_plugin/net_plugin.cpp [hunk 12]
```diff
@@ -1435,6 +1476,18 @@ namespace eosio {
...
+      if(participant_name_ && !participating_) {
+         const bool ignore = std::holds_alternative<notice_message>(m) ||
+                             std::holds_alternative<signed_block_v0>(m) ||
+                             std::holds_alternative<packed_transaction_v0>(m) ||
+                             std::holds_alternative<signed_block>(m) ||
+                             std::holds_alternative<trx_message_v1>(m);
+         if(ignore) {
+            return;
+         }
+      }
```
[the remark this answers is not in the document]
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:1490]: Added

File: tests/CMakeLists.txt (hunk 2 above)
Comment (rusty-block-one) [reply tests/CMakeLists.txt:8]: When I initially developed the unit test, I placed it in the tests directory.  When we decided to create a test directory under net_plugin for unit tests, I did not backtrack and remove the change.  The file has been modified back to the original.

File: plugins/net_plugin/net_plugin.cpp [hunk 13]
```diff
@@ -3428,10 +3481,51 @@ namespace eosio {
       }
    }
 
+   // called from any thread
+   void net_plugin_impl::update_security_group(const block_state_ptr& bs) {
+      // update cache
```
[the remark this answers is not in the document]
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:3486]: Now using connections_mtx

File: plugins/net_plugin/net_plugin.cpp [hunk 14]
```diff
@@ -3428,10 +3481,51 @@ namespace eosio {
...
+               added_connections.push_back(connection);
+            }
+         }
+         else {
+            connection->now_syncing();
+         }
+         return true;
+      };
+
+      for_each_connection(do_update);
```
[the remark this answers is not in the document]
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:3515]: changed to loop on connections instead of calling for_each_connection

File: plugins/net_plugin/net_plugin.cpp [hunk 15]
```diff
@@ -254,8 +256,9 @@ namespace eosio {
       bool                                  use_socket_read_watermark = false;
       /** @} */
 
-      mutable std::shared_mutex             connections_mtx;
+      mutable std::shared_mutex             connections_mtx; // protects both connections and security_group
```
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:259]: As implemented `connections_mtx` does not protect the `security_group`. It looks to me like `security_group` is only accessed and modified via the application thread.

File: plugins/net_plugin/net_plugin.cpp [hunk 16]
```diff
@@ -3428,10 +3470,53 @@ namespace eosio {
...
+            connection->set_participating(false);
+         }
+      };
+
+      // send handshake when added to group
+      //    - connection mutex no longer needd
+      //
+      connection_guard.unlock();
+      for(auto& connection : added_connections) {
+         connection->send_handshake();
```
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:3511]: This just does a post on a strand. I don't see the need to pull this out and send after releasing the mutex. Just call send_handshake() above so that `added_connections` is not needed.

File: plugins/net_plugin/net_plugin.cpp (hunk 15 above)
Comment (brianjohnson5972) [reply plugins/net_plugin/net_plugin.cpp:259]: It will need to be checked by the connection itself after the certificates are exchanged, to determine if the incoming participant is in the security group and can receive blocks or not.  I thought it would be simpler for it to be covered by the connections mutex rather than having a separate one and having to make sure that they are taken in the correct order.

Comment (heifner) [reply plugins/net_plugin/net_plugin.cpp:259]: I don't think it needs a mutex as it is only accessed via the application thread, right?

Comment (brianjohnson5972) [reply plugins/net_plugin/net_plugin.cpp:259]: The TLS connection portion isn't implemented yet. The connection will have an extra handshake where the participant will be identified and the connection will then have to search in the security group to see if the participant is in or out of the group, and that won't be on the application thread, it will be on one of the net_plugin thread pool threads.

File: plugins/net_plugin/net_plugin.cpp [hunk 17]
```diff
@@ -3484,38 +3473,40 @@ namespace eosio {
...
-      // update connection
+      // update connections
+      //    - connection_mtx still needed
       //
-      auto do_update = [&](auto& connection) {
+      for(auto& connection : connections) {
          const auto& participant = connection->participant_name();
          if(!participant) {
-            return true;
+            return;
```
Comment (brianjohnson5972) [inline plugins/net_plugin/net_plugin.cpp:3493]: this should be continue instead of return

File: plugins/net_plugin/net_plugin.cpp [hunk 18]
```diff
@@ -3428,10 +3470,53 @@ namespace eosio {
       }
    }
 
+   // called from any thread
+   void net_plugin_impl::update_security_group(const block_state_ptr& bs) {
+      // update cache
+      //    - connection_mtx is needed
+      //
+      auto& update = bs->get_security_group_info();
+      std::shared_lock<std::shared_mutex> connection_guard(connections_mtx);
```
Comment (heifner) [inline plugins/net_plugin/net_plugin.cpp:3479]: This needs to be a `lock_guard` as it is modifying state.

File: plugins/net_plugin/net_plugin.cpp (hunk 15 above)
Comment (heifner) [reply plugins/net_plugin/net_plugin.cpp:259]: In that case it needs to be a lock_guard not a shared_lock of the connections_mtx.

File: plugins/net_plugin/net_plugin.cpp [hunk 19]
```diff
@@ -1446,6 +1485,9 @@ namespace eosio {
    }
 
    void connection::enqueue_block( const signed_block_ptr& b, bool to_sync_queue) {
+      if(!is_participating()) {
```
Comment (brianjohnson5972) [inline plugins/net_plugin/net_plugin.cpp:1488]: Need similar code (except with continue instead of return) also in dispatch_manager::bcast_block, after the if (!sb) block that sends the go away message, before block is sent to enque_buffer. Also in dispatch_manager::bcast_transaction, need to change:

if( cp->is_blocks_only_connection() || !cp->current() )

into:

if( cp->is_blocks_only_connection() || !cp->current() || !cp->is_participating() )

I believe this covers everything, but double check me that the "Need to check participating flag before calling enque_buffer for any transaction, block, or notice_messages for any connection." requirement has been met.

File: plugins/net_plugin/net_plugin.cpp [hunk 20]
```diff
@@ -3428,10 +3470,53 @@ namespace eosio {
...
+
+      // update connections
+      //    - connection_mtx still needed
+      //
+      for(auto& connection : connections) {
+         const auto& participant = connection->participant_name();
+         if(!participant) {
+            return;
+         }
+         if(security_group.is_in_security_group(participant.value())) {
```
Comment (brianjohnson5972) [inline plugins/net_plugin/net_plugin.cpp:3495]: Need to change sec;urity_group_manager::is_in_security_group to always return true if cache_.empty().  This will allow this logic here to work if we ever remove all participants in the security group and it allows Dmytro's code to only have to call that method to determine what he should set participating_ to.

**Review by brianjohnson5972 — CHANGES_REQUESTED** [review]

File: plugins/net_plugin/net_plugin.cpp (hunk 17 above)
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:3493]: changed to continue

File: plugins/net_plugin/net_plugin.cpp (hunk 16 above)
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:3511]: Removed added_connections and now call send_handsahke in update loop

File: plugins/net_plugin/net_plugin.cpp (hunk 15 above)
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:259]: Changed to lock_guard

File: plugins/net_plugin/net_plugin.cpp (hunk 20 above)
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:3495]: Modified is_in_security_group to also return true when the cache_ is empty

File: plugins/net_plugin/net_plugin.cpp (hunk 18 above)
Comment (heifner) [reply plugins/net_plugin/net_plugin.cpp:3479]: This is called on every `on_accepted_block` which means we now have a mutex on the critical performance path. Can we re-design this so that we don't need to acquire a mutex on every block.

File: plugins/net_plugin/net_plugin.cpp (hunk 19 above)
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:1488]: In dispatch_manager::bcast_block, the "if(!sb)" block is in the lambda passed to for_each_block_connection.  So "return true" actually results in a continue to the next connection.

I did a search on all references to enqueue_buffer and each has a preceding call to is_participating() in the code to bypass the call to enqueue_buffer.

File: plugins/net_plugin/net_plugin.cpp [hunk 21]
```diff
@@ -2237,6 +2237,10 @@ namespace eosio {
             return true;
          }
 
+         if(!cp->is_participating()) {
```
Comment (brianjohnson5972) [inline plugins/net_plugin/net_plugin.cpp:2240]: Sorry, this block should go before line 2231 to avoid bothering to create the send_buffer

File: plugins/net_plugin/net_plugin.cpp [hunk 22]
```diff
@@ -3428,10 +3474,40 @@ namespace eosio {
       }
    }
 
+   // called from any thread
+   void net_plugin_impl::update_security_group(const block_state_ptr& bs) {
+      // update cache
+      //
+      auto& update = bs->get_security_group_info();
+      std::lock_guard<std::shared_mutex> connection_guard(connections_mtx);
```
Comment (brianjohnson5972) [inline plugins/net_plugin/net_plugin.cpp:3482]: still need to add something like:
// NOTE: since this is the only thread that ever updates security_group.version it is safe to check this before taking the lock
if (security_group.version == update.version) {
   return;
}

just before taking the lock to avoid taking the lock on every block when we only need to when the version changed.

**Review by brianjohnson5972 — CHANGES_REQUESTED** [review]

File: plugins/net_plugin/net_plugin.cpp (hunk 21 above)
Comment (rusty-block-one) [reply plugins/net_plugin/net_plugin.cpp:2240]: moved the test prior to creating send_buffer

File: plugins/net_plugin/test/security_group_manager_tests.cpp [hunk 23]
```diff
@@ -37,6 +37,16 @@ BOOST_AUTO_TEST_CASE(test_initial_population) {
       BOOST_REQUIRE(manager.is_in_security_group(participant));
    }
 }
+
+BOOST_AUTO_TEST_CASE(test_version) {
+   eosio::security_group_manager manager;
+   BOOST_REQUIRE(manager.current_version() == 0);
+
+      auto populate = create_list({ 1, 2, 3, 4, 5, 6});
```
Comment (brianjohnson5972) [inline plugins/net_plugin/test/security_group_manager_tests.cpp:45]: fix the indentation here

**Review by brianjohnson5972 — CHANGES_REQUESTED** [review]
Indentation error is the last thing and then we can merge

File: plugins/net_plugin/test/security_group_manager_tests.cpp (hunk 23 above)
Comment (rusty-block-one) [reply plugins/net_plugin/test/security_group_manager_tests.cpp]: indentation fixed

**Review by brianjohnson5972 — APPROVED** [review]
